site/content/en/concepts/architecture/_index.md

```diff
@@ -0,0 +1,156 @@
+configuration outside of the Kubernetes control-plane
+(i.e. operating against git + files rather than against the apiserver)
+
+This design provides a GitOps / resource + controller hybrid model which:
+
+- Aligns kpt with the tools and APIs provided by the Kubernetes project.
+- Enables a framework for loosely coupling tools within the Kubernetes ecosystem
+  (e.g. kpt, kustomize, kubectl, kubeval).
+- Minimizes fragmentation by building on top of the primitives provided by the Kubernetes project.
+  - e.g the Kubernetes examples are stored as resource configuration in git
+
+## Design Principles
+
+### Resource oriented
+
+Operations are defined by creating, modifying and deleting Kubernetes resources.  Imperative
+operations are implemented as writes to declarative resources.
+
+#### kpt vs kubernetes:
+
+- kpt: resource configuration read and written locally (file operations)
+- kubernetes: resources read and written through the apiserver (http operations)
+
+### Symmetric Input / Output
+
+Operations read configuration and modify it.  The kpt inputs and outputs should be symmetric.
+It should be possible to pipe kpt operations together, and it should be possible to write
+the output of a kpt operation back to its source (updating it in-place).
+
+#### kpt vs kubernetes:
+
+- kpt: resource configuration is read, modified and written back to its source
+  - resources may be updated using 3-way merge (kpt pkg update)
+- kubernetes: resources are read, modified and written back to the apiserver
+  - resources may be updated using 3-way merge (kubectl apply)
+
+### Schema driven
+
+Type or object specific logic should NOT be built into the tool.
+Static resource modifications (e.g. *set*) should be configured using type or object
+metadata (e.g. schema).
+
+OpenAPI is used for resource schema.  Tools may support their own OpenAPI
+extensions which should co-exist with extensions owned by other tools.
+
+Support for new types should be introduced through new OpenAPI definitions rather than
+changes to the tool itself.
+
+- Static per-type and per-object resource transformations should use OpenAPI to tell the
+  tool how to modify a given object
+  - e.g. where to set `image` for `set image` is defined in OpenAPI rather than hard-coded
+  
+- Configuration for individual objects / resources may define custom OpenAPI definitions for
+  that specific instance
+  - e.g. an nginx Deployment's `image` may be restricted to the regular expression `^nginx:.*$`
+
+#### kpt vs kubernetes:
+
+- kpt: OpenAPI read from multiple sources -- can also be inlined into individual
+  configuration objects as comments
+- kubernetes: OpenAPI read from apiserver
+
+### Resource + Controller architecture
+
+The desired state of the system should be expressed as declarative, static data structures
+rather than as executable or interpreted code (DSLs and templates are code).  
+
+Dynamic logic should be built into programs which read configuration as input, modify it
```

Comment:
May want to discuss two categories of such programs: 1) Common operations built into kpt CLI 2) Programs using the functions extendability mechanism.

Reply:
talked about this in composability